libs/cardano-ledger-core/internal/Cardano/Ledger/Internal/Definition/Era.hs

```diff
@@ -10,16 +22,162 @@ module Cardano.Ledger.Internal.Definition.Era (
   DijkstraEra,
 ) where
 
+import Cardano.Ledger.Binary (MaxVersion, MinVersion)
+import Data.Kind (Type)
+import Data.Typeable (Typeable)
+import GHC.Exts (Proxy#, proxy#)
+import GHC.TypeLits
+
+--------------------------------------------------------------------------------
+-- Era
+--------------------------------------------------------------------------------
+
+class
+  ( Typeable era
+  , KnownNat (ProtVerLow era)
+  , KnownNat (ProtVerHigh era)
+  , ProtVerLow era <= ProtVerHigh era
+  , MinVersion <= ProtVerLow era
+  , MinVersion <= ProtVerHigh era
+  , -- We need to make sure that there is never a case that MaxVersion equals to the highest
+    -- protocol version for the era, otherwise we can never upgrade to the next version:
+    CmpNat (ProtVerLow era) MaxVersion ~ 'LT
+  , CmpNat (ProtVerHigh era) MaxVersion ~ 'LT
+  ) =>
+  Era era
+  where
+  type EraName era :: Symbol
+
+  -- | Map an era to its predecessor.
+  --
+  -- For example:
+  --
+  -- > type instance PreviousEra (AllegraEra c) = ShelleyEra c
+  type PreviousEra era = (r :: Type) | r -> era
+
+  -- | Lowest major protocol version for this era
+  type ProtVerLow era :: Nat
+
+  -- | Highest major protocol version for this era. By default se to `ProtVerLow`
+  type ProtVerHigh era :: Nat
+
+  type ProtVerHigh era = ProtVerLow era
+
+  -- | Textual name of the current era.
+  --
+  -- Designed to be used with @TypeApplications@:
+  --
+  -- >>> eraName @ByronEra
+  -- "Byron"
+  eraName :: String
+  default eraName :: KnownSymbol (EraName era) => String
+  eraName = symbolVal' (proxy# :: Proxy# (EraName era))
+
+-- | This is a non-existent era and is defined for satisfying the `PreviousEra` type family injectivity
+data VoidEra
+
+-- | This is the era that preceded Shelley era. It cannot have any other class instances,
+-- except for `Era` type class.
+data ByronEra
+
+instance Era ByronEra where
+  type EraName ByronEra = "Byron"
+  type PreviousEra ByronEra = VoidEra
+  type ProtVerLow ByronEra = 0
+  type ProtVerHigh ByronEra = 1
+
 data ShelleyEra
 
+instance Era ShelleyEra where
+  type EraName ShelleyEra = "Shelley"
+  type PreviousEra ShelleyEra = ByronEra
+  type ProtVerLow ShelleyEra = 2
+
 data AllegraEra
 
+instance Era AllegraEra where
+  type EraName AllegraEra = "Allegra"
+  type PreviousEra AllegraEra = ShelleyEra
+  type ProtVerLow AllegraEra = 3
+
 data MaryEra
 
+instance Era MaryEra where
+  type EraName MaryEra = "Mary"
+  type PreviousEra MaryEra = AllegraEra
+  type ProtVerLow MaryEra = 4
```

Comment:
It seems that we don't need to also specify `ProtVerHigh` for shelley, allegra and mary. But I don't understand why. 😅

Reply:
See a few lines above, it has a default definition:
```
  type ProtVerHigh era :: Nat
  type ProtVerHigh era = ProtVerLow era
```